[feat] Home API 기능 구현 완료

dev-msj opened the pull request:
## 작업 내용
DB의 USER_CATEGORY에서 요청 UID에 해당하는 데이터 조회
Redis에서 요청 책 추천 타입에 해당하는 데이터를 조회
Redis가 Empty 일 때, DB에서 Book Suggestion 데이터를 가져와 Redis에 Push
UserCategory에 해당하는 추천 책 목록 반환
요청 내용 캐싱
간단한 코드 정리

## 이슈 번호

#9 홈 API 구현

dev-msj committed:
docs : Home API에 필요한 기능 정의

dev-msj committed:
docs : Home API 기능 정의

dev-msj committed:
chore : Spring Security 비활성화 및 Mapper 경로 수정

dev-msj committed:
feat : Home API 개발 완료

DB의 USER_CATEGORY에서 요청 UID에 해당하는 데이터 조회
Redis에서 요청 책 추천 타입에 해당하는 데이터를 조회
Redis가 Empty 일 때, DB에서 Book Suggestion 데이터를 가져와 Redis에 Push
UserCategory에 해당하는 추천 책 목록 반환
요청 내용 캐싱

dev-msj committed:
refactor : BookSuggestionConverter 코드 정리

BookSuggestionParser에서 BookSuggestionConverter로 변경
일부 함수, 변수 이름 변경
메서드가 SRP에 부합하도록 변경

dev-msj committed:
style : 사용되지 않는 객체 및 import 삭제

dev-msj committed:
docs : 마크다운 오류 수정

dev-msj committed:
style : convert 사용하지 않고 parse만 사용하도록 네이밍 컨벤션 적용

dev-msj committed:
refactor : JsonUtils.parseValueToBookList를 BookSuggestParser로 이동

dev-msj committed:
style : model 객체 이름에서 DTO 삭제

dev-msj committed:
style : 코드 정리

dev-msj committed:
refactor : RedisService null 예외처리 추가

dev-msj committed:
refactor : suggestion type을 enum으로 관리하도록 변경

dev-msj committed:
test : SuggestionType Enum 테스트 코드 추가

dev-msj committed:
refactor : suggestion 캐싱키 변경

Custom Key Generator인 OasisKeyGenerator 생성.
method_param0_param1...의 형식으로 key를 구성하여 추후 key 추적이 가능하도록 변경.

dev-msj committed:
test : HomeService의 suggestion 메서드 결과 캐싱 확인 테스트 추가

dev-msj committed:
chore : 패키지 경로 수정

enum 클래스들을 constant 경로에서 관리하도록 변경
OasisKeyGenerator를 config 하위 경로로 이동

dev-msj committed:
refactor : RedisKey를 enum으로 관리하도록 변경

dev-msj committed:
refactor : userCategory가 없을 경우 전체 카테고리에 해당하는 데이터를 반환하도록 변경

dev-msj committed:
refactor : redis key가 존재하지 않을 경우 RuntimeException을 던지도록 변경

dev-msj committed:
refactor : HomeService에서 json 관련 로직들을 JsonUtils에서 처리하도록 변경

dev-msj committed:
refactor : JsonUtils에서 JsonNode에 대한 Null 예외처리 추가

Key, Value 쌍의 JsonArray 형태를 가진 JsonNode에서 Key가 존재하지 않을 때
Runtime Exception을 던지도록 변경

dev-msj committed:
fix : HomeService에서 Stream을 List로 변경하는 코드 오류 수정

jinspark-lab commented on src/main/java/com/flab/oasis/controller/home/HomeController.java:
```diff
@@ -0,0 +1,25 @@
+    private final HomeService homeService;
+
+    @PostMapping("/suggestion")
+    public List<Book> bookSuggestion(@RequestBody Map<String, String> map) {
```
JSON 을 받아서 파싱하는 대신 Map 으로 받으려 하시는 건가요?

dev-msj replied:
RequestBody의 경우에 Model과 동일한 구조를 가지는 json에 대해서는 자동 형변환을 해주고, 그렇지 않을 경우에는 Map을 활용해서 받는 것으로 알고 있어서 위와 같이 작성했습니다.

말씀해주신 내용처럼 Json으로 받아서 파싱하는 방법에 대해서도 확인해보도록 하겠습니다!

jinspark-lab replied:
넵.. 이렇게 하시면 어떤 팀원도 Request 를 예측할 수가 없습니다.

dev-msj replied:
다른 누군가가 봤을 때 예측이 가능하게 작성하는 것이 중요한데, 그 부분을 고려하지 못했네요..
해당 부분은 데이터를 받아오는 DTO를 따로 만들어서 적용하도록 하겠습니다!

dev-msj replied:
해당 내용 반영했습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/model/Book.java:
```diff
@@ -17,29 +22,9 @@ public class Book {
-                getPublishDate(), getImageUrl(), getDescription(), getCategoryId()
-        );
-    }
+    @Transient
```
이전 피드백과 동일합니다

dev-msj replied:
해당 내용 반영했습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/service/home/BookSuggestionParser.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+class BookSuggestionParser {
```
구조 수정 리뷰 내용 동일합니다

dev-msj replied:
일단 나름대로 구조 수정해봤습니다!
다만 제가 작업한 내용이 thomas님께서 피드백 해주신 내용에 부합하는지 확신이 안서서 다시 한 번 리뷰 부탁드립니다!🙏

jinspark-lab replied:
전보다 나아졌지만, 더 나은 방법이 있습니다. 이 부분 관련해서 전체 커멘트로 한번 드릴게요!

dev-msj replied:
service가 redis 데이터를 다루는 모델만 바라보게 했고, 해당 모델에서 db 데이터를 컨트롤할 수 있도록 다시 수정했습니다!

jinspark-lab commented on src/test/java/com/flab/oasis/home/SuggestionTypeTest.java:
```diff
@@ -0,0 +1,14 @@
+    void suggestionTypeTest() {
+        String type = "bestSeller";
+        SuggestionType suggestionType = SuggestionType.valueOf(type.toUpperCase());
+        Assertions.assertEquals(type, suggestionType.getSuggestionType());
```
테스트 의도를 잘 모르겠습니다.

dev-msj replied:
제가 확인하기 위해서 작성했던 테스트 코드인데, 이런 개인적인 레벨의 코드는 팀에게 공유될 필요가 없다는 것을 생각하지 못했네요!
해당 코드는 삭제하고, 앞으로 이런 레벨의 테스트는 커밋에서 제외하도록 하겠습니다!

jinspark-lab replied:
그렇다기보다는.. String 으로 Enum 을 가져오고, 리턴된 값을 다시 원본 값이랑 비교하면 항상 성공할 테스트 같습니다. 휴먼 에러 테스팅 용도신가요? 휴먼에러라면 이렇게 테스트해서 잡는건 이상합니다. 자동화로 검사해야죠.

dev-msj replied:
처음 의도했던 내용은 동일한 값을 가진 enum 객체가 항상 같은 hashCode를 가지는지 테스트 하는 것이었습니다.
하지만 해당 코드는 그저 동일한 String 값 비교이기 때문에 항상 성공할 테스트로 의미 없는 테스트가 된 것 같습니다. 또한 본래의 목적과도 부합하지 않는 테스트이기도 하네요...

jinspark-lab reviewed (changes requested):
수업시간에 말씀드린 구조 부분은 따로 피드백 남기지 않았습니다.

dev-msj committed:
style : Book Model에서 JsonIgnore를 사용하고 있기 때문에 Trasient는 삭제

dev-msj committed:
refactor : HomeMapper의 List 반환값들을 일급 컬렉션으로 wrapping

일급 컬렉션 안에서 관련 비즈니스 로직들이 처리되도록 변경

dev-msj committed:
refactor : BookSuggestion에서suggestionType의 타입을 String에서 SuggestionType으로 변경

dev-msj committed:
refactor : redis 결과값을 Json Model로 맵핑해서 처리되도록 변경

RedisService에서 Json 모델에 맵핑하는 로직 추가
RedisKey에서 적절한 모델에 맵핑되도록 expression 추가
BookSuggestionJson에서 json paring이 이뤄지도록 변경
HomeService 코드 정리

dev-msj committed:
style : 필요없는 테스트 코드 삭제

jinspark-lab reviewed (commented):
혼란에 빠지신 것 같아서 전체적으로 레퍼런스를 좀 드려야할 것 같습니다. Spring 의 정석적인 코드를 좀 더 보시면 도움이 될 것 같습니다.

참고하실만한 소스 레퍼런스 : 
- 수강생 프로젝트지만 상당히 실무스러운 코드입니다 => https://github.com/f-lab-edu/food-delivery
- 제 토이프로젝트이고 레거시입니다만 도움은 되실겁니다. 여긴 Redis 는 없습니다. => https://github.com/jinspark-lab/jinspark-lab

개념 정리 : 
- Spring MVC : https://techhan.github.io/study/spring-08/
- Spring 기본 구조 : https://velog.io/@jybin96/Controller-Service-Repository-%EA%B0%80-%EB%AC%B4%EC%97%87%EC%9D%BC%EA%B9%8C

Request - Response 등을 각각 객체로 관리하시고 JsonMapper 를 붙이셔서 각 계층에서 파싱하게 하시면 많은 부분이 해결되십니다. 직접 파서를 구축해서 구현하는 방식도 있지만 전혀 그러실 이유가 없습니다.

위의 프로젝트랑 글들을 보시면서 좀 더 이해하시면 도움이 될겁니다.

dev-msj committed:
refactor : redis, db 관련된 데이터 제어를 model에서 이뤄지도록 변경

dev-msj committed:
fix : 쿼리 결과가 enum에 매칭되지 않는 문제 해결

dev-msj committed:
refactor : SuggestionType 관련 데이터는 대문자로 활용하도록 변경

dev-msj committed:
fix : redisTemplate 결과가 타입과 매칭되지 않는 문제 해결

dev-msj committed:
refactor : RequestBody를 BookSuggestionRequest 모델로 받도록 수정

dev-msj committed:
style : BookSuggestionRedis method 순서 변경

jinspark-lab commented on src/main/java/com/flab/oasis/model/home/BookSuggestionRedis.java:
```diff
@@ -0,0 +1,60 @@
+
+@Service
+@AllArgsConstructor
+public class BookSuggestionRedis {
```
Model 안에 Service 가 있습니다.

dev-msj replied:
repository 패키지로 보내고 Repository 어노테이션으로 변경했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/home/BookSuggestionRedis.java:
```diff
@@ -0,0 +1,60 @@
+@Service
+@AllArgsConstructor
+public class BookSuggestionRedis {
+    private final StringRedisTemplate redisTemplate;
```
위와 동일하고, Model 안에서 Service 로직을 수행하고 있습니다.

dev-msj replied:
Repository에서 Service로 판단되는 부분은 BookSuggestionService로 옮겼습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/home/BookSuggestionRedis.java:
```diff
@@ -0,0 +1,60 @@
+        return Optional.ofNullable(hashOperations.get(key.name(), field)).orElseGet(() -> {
+            pushNewBookSuggestion(getBookSuggestionHashFieldValue());
+
+            return getBookSuggestionFromRedis(key, field);
```
굉장히 위험한 코드입니다. 이 부분은 왜 위험한지 잘 생각해보셔요.

dev-msj replied:
redis 관련 이슈가 발생하여 데이터를 가져올 수 없을 때 무한 루프를 돌아버리겠네요...

jinspark-lab replied:
하나 더 있습니다. 코드 관리의 문제에요. 로직을 한번 잘 고민해보셔요.

dev-msj replied:
해당 로직은 db에서 꺼내온 데이터에서 직접 반환하도록 변경했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/home/BookSuggestionRedis.java:
```diff
@@ -0,0 +1,60 @@
+
+    private List<Integer> getUserCategoryByUid(String uid) {
+        return new UserCategoryCollection(homeMapper.findUserCategoryByUid(uid)).getCategoryIdList();
+    }
```
구조가 잘못되어있습니다.

dev-msj replied:
해당 로직은 리팩토링 과정에서 삭제되었습니다!

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -0,0 +1,45 @@
+
+        Cache cache = cacheManager.getCache("homeCache");
+
+        Assertions.assertNotNull(cache);
```
Not Null 로 테스트하기 적합하지 않아보입니다. 아래 처럼 값을 꺼내서 비교하는게 더 확실해보입니다.
좀 더 좋은 코드는 @Before , @After 를 달아서 Before 안에서 Insert 를 쭉 해놓고 테스트 돌린 이후 After 안에서 전부 삭제해주는 겁니다.

jinspark-lab replied:
TBD

jinspark-lab commented on src/main/java/com/flab/oasis/model/home/BookSuggestionCollection.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private String getSuggestionType(Map.Entry<SuggestionType, List<BookSuggestion>> entry) {
+        return entry.getKey().name();
```
NPE 가능성이 높아보입니다.

dev-msj replied:
해당 로직은 리팩토링 과정에서 삭제되었습니다!

jinspark-lab commented on src/main/java/com/flab/oasis/controller/home/HomeController.java:
```diff
@@ -0,0 +1,26 @@
+public class HomeController {
+    private final HomeService homeService;
+
+    @PostMapping("/suggestion")
```
Post 로 설계하신 이유가 궁금합니다. 바꾸라는 얘기는 아니고, 의도가 있으신지 체크하고 싶습니다.

dev-msj replied:
추후 jwt를 사용하게 되면 토큰을 같이 보내야한다는 생각때문에 위과 같이 사용했습니다.

dev-msj committed:
refactor : json string 파싱을 BookSuggestionJson 모델에서 하도록 변경

jinspark-lab reviewed (changes requested):
말씀드린 Request / Response 부분은 나아졌는데, 죄송합니다만 아직 부족합니다...
모델에 서비스 클래스를 넣으신건 실수이신거죠? 만약 실수가 아니라면 학습이 더 필요해보입니다. 보내드린 링크 정독해보시기를 권장드려요.
전체적으로 코드들이 너무 무겁습니다. 모델에 로직이 많고, 솔직히 왜 DTO 가 아닌 Map, String, Integer 들을 이렇게 많이 전달되고 주고받는지 잘 모르겠습니다.
수업시간에 방향을 다시 한번 같이 논의해보시죠. :)

dev-msj committed:
fix : JsonUtils를 사용하던 코드 에러 수정

dev-msj committed:
refactor : RedisTemplate을 <String, Object>형태로 사용하도록 변경

dev-msj committed:
style : Book에서 categoryId의 JsonIgnore 삭제

dev-msj committed:
refactor : BookSuggestion에서 suggestionType을 String으로 변경

dev-msj committed:
refactor : BookSuggestionRepository에서 redis 및 db 데이터를 다루도록 변경

dev-msj committed:
refactor : BookSuggestionRepository에서 Service 로직 분리

jinspark-lab reviewed (commented):
수업시간 피드백 바탕으로 수정해주시면 반영하겠습니다.

dev-msj committed:
refactor : BookSuggestionRepository에서 RedisKey가 변경될 일이 없으므로 멤버 변수로 사용하도록 변경

dev-msj committed:
refactor : suggestionType이 SuggestionType enum으로 관리되도록 변경

dev-msj committed:
style : repository를 service 하위 패키지로 이동

repository 상에 service적인 로직이 필요하므로 service 패키지에서 관리하도록 변경

jinspark-lab commented on src/main/java/com/flab/oasis/service/home/BookSuggestionRepository.java:
```diff
@@ -0,0 +1,50 @@
+
+@Repository
+@RequiredArgsConstructor
+public class BookSuggestionRepository {
```
Service 하위에 Repository 는 왜 옮기신건가요?

dev-msj replied:
수업시간 때 repository가 serivce적인 로직을 품고 있어서 service 하위로 들어가도 무관할 것 같다는 피드백을 해주셔서 옮겼었습니다.
혹시 제가 잘못 이해했던 것일까요?

jinspark-lab replied:
넵. 잘못 이해하셨습니다. Repository 가 아니라 Service 가 되어도 무방하다는 뜻이었습니다.

dev-msj replied:
아...해당 부분에 대해서는 spring mvc에 대한 공부를 다시 하고 수정해보도록 하겠습니다!

dev-msj replied:
BookSuggestionRepository를 repository 패키지로 옮기고 Service 어노테이션을 할당하도록 변경했습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/service/home/BookSuggestionRepository.java:
```diff
@@ -0,0 +1,50 @@
+    private final ObjectMapper objectMapper;
+    private final HomeMapper homeMapper;
+
+    private final RedisKey key = RedisKey.HOME;
```
현구조에서는 이렇게 해도 무방하지만 결국에는 비효율적인 구조라는건 이해하셔야합니다. 너무 시간 오래 끌 수 없으니 일단 나중에 리팩토링하는 걸로 하고 넘어가겠습니다.

dev-msj replied:
SuggestionType을 key로 사용하도록 변경했습니다.
사용하던 structure도 hash에서 string으로 변경했습니다!

jinspark-lab replied:
그 문제는 아닙니다. Redis 사용하는 구조 자체를 템플릿화할 수 있다는 뜻입니다. 이 건은 지금 해결할 수 없으니 넘어가겠습니다.

dev-msj replied:
네 알겠습니다!

jinspark-lab reviewed (commented):
Spring MVC 구조 공부하시고 커밋 수정해주세요. 수정 이후 타임라인 상 PR 은 반영되겠지만, 리팩토링을 별도 이슈로 빼겠습니다. 
그리고 커밋 메시지 이메일 알림으로 소통하지 마시고, 반영하셨으면 PR 에 커멘트로 변경내용 설명을 자세히 달아주시는게 좋습니다. 실무에서는 말하기 전까지 보통 안본다고 생각하시면 됩니다.

dev-msj committed:
refactor : suggestionType을 redis의 키로 사용하도록 변경

기존에 hash structure를 string structure로 변경

dev-msj committed:
refactor : redis에서 key를 string이 아닌 SuggestionType enum으로 사용하도록 변경

dev-msj committed:
style : BookSuggestionRepository를 repository로 이동.

이 클래스는 Repository의 역할을 하나 Service적인 로직이 존재한다.
그래서 repository 패키지에서 관리하도록 하고, Repository 어노테이션을 Service로 변경한다.

jinspark-lab commented on src/main/java/com/flab/oasis/repository/home/BookSuggestionRepository.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+@Service
```
@Repository 로 바꿔주세요.

jinspark-lab commented on src/main/java/com/flab/oasis/model/home/BookSuggestionRequest.java:
```diff
@@ -0,0 +1,12 @@
+
+@Getter
+@Setter
+public class BookSuggestionRequest {
```
통신이 되는 객체의 경우 Serializable 을 구현해주세요.

jinspark-lab reviewed (commented):
커멘트 확인해주시고 Merge 하겠습니다

dev-msj committed:
refactor : 통신이 되는 객체이므로 Serializable 상속

dev-msj committed:
refactor : BookSuggestionRepository에서 Service를 Repository로 변경
